uwblueprint/jump-math
# fix: Refactor test backend to support images

## Notion ticket link

[Refactor test backend to support images](https://www.notion.so/uwblueprintexecs/Refactor-test-backend-to-support-images-5b7afdfa5041408fb3ceb643a21db893?pvs=4)



## Implementation description
- **GRAPHQL**: Rename using new types
- **MODEL**: Moved question related interfaces / types into new files (questionTypes.ts and questionMetadataTypes.ts)
- **UNIT TESTS**: Refactored tests and use new variables
- **SERVICE**: 
   - upload images before creating / updating test
   - hydrate before returning
- **TEST UTILS**: Updated mock questions to include image component
- **TYPES**: add new question component types for GraphQL request (GraphQLQuestionComponent), service request (QuestionComponentRequest), and model / service response (QuestionComponent)


## Steps to test
1. Unit tests
2. Ran every existing query / mutation in GraphQL Playground to make sure it was not broken 👍🏻 - i don't think its possible to mock up a file upload in playground so i didn't actually test w/ an image. But all existing code should not be broken, and I will test directly later when FE is built.

(example)
<img width="1308" alt="image" src="https://user-images.githubusercontent.com/69697180/231839125-ed4bcd2b-8006-4857-b097-3110f1d3eb9a.png">




## What should reviewers focus on?
- Main files to review:
   - `backend/services/implementations/testService.ts`
   - `backend/types/questionMetadataTypes.ts`
   - `backend/types/questionTypes.ts`
- Overall code quality

FYI Notes
- I noticed duplicateTest doesn’t validate for status field
    - Complicated bc unarchiveTest also calls duplicateTest
    - FE should be enough validation
- TODO: delete uploaded images on fail (catch block) https://www.notion.so/uwblueprintexecs/Rollback-Image-Upload-on-Failure-c5d5900235d744448b2efcf5652798f2?pvs=4


## Checklist
